samples/matter/common/src/bridge/bridge_manager.cpp

```diff
@@ -29,50 +40,66 @@ void BridgeManager::Init()
 {
-	VerifyOrReturnError(devices && dataProvider, CHIP_ERROR_INTERNAL);
+	chip::Optional<uint8_t> indexes[deviceListSize];
+	chip::Optional<uint8_t> *indexPtrs[deviceListSize];
+	uint16_t endpoints[deviceListSize];
 
-	dataProvider->Init();
+	VerifyOrReturnError(devicesPairIndexes, CHIP_ERROR_INTERNAL);
 
-	/* Wrap input data into unique_ptr to avoid memory leakage in case any error occurs. */
-	Platform::UniquePtr<MatterBridgedDevice> devicesPtr[deviceListSize];
-	for (auto i = 0; i < deviceListSize; ++i) {
-		devicesPtr[i] = std::move(Platform::UniquePtr<MatterBridgedDevice>(devices[i]));
+	for (auto i = 0; i < deviceListSize; i++) {
+		indexPtrs[i] = &indexes[i];
 	}
-	Platform::UniquePtr<BridgedDeviceDataProvider> dataProviderPtr(dataProvider);
 
-	/* Maximum number of Matter bridged devices is controlled inside mDevicesMap,
-	   but the data providers may be created independently, so let's ensure we do not
-	   violate the maximum number of supported instances. */
-	VerifyOrReturnError(mDevicesMap.FreeSlots() >= deviceListSize, CHIP_ERROR_NO_MEMORY,
-			    LOG_ERR("Not enough free slots in the map"));
-	VerifyOrReturnError(mNumberOfProviders + 1 <= kMaxDataProviders, CHIP_ERROR_NO_MEMORY,
-			    LOG_ERR("Maximum number of providers exceeded"));
-	mNumberOfProviders++;
+	return AddBridgedDevices(devices, dataProvider, deviceListSize, devicesPairIndexes, endpoints, indexPtrs);
+}
 
-	bool cumulativeStatus{ true };
-	bool status{ true };
-	for (auto i = 0; i < deviceListSize; ++i) {
-		MatterBridgedDevice *device = devicesPtr[i].get();
-		status = AddSingleDevice(device, dataProviderPtr.get());
-		if (status) {
-			devicesPtr[i].release();
-		}
-		cumulativeStatus &= status;
+CHIP_ERROR BridgeManager::AddBridgedDevices(MatterBridgedDevice *devices[], BridgedDeviceDataProvider *dataProvider,
+					    uint8_t deviceListSize, uint8_t *devicesPairIndexes[],
+					    uint16_t endpointIds[])
+{
+	chip::Optional<uint8_t> indexes[deviceListSize];
+	chip::Optional<uint8_t> *indexPtrs[deviceListSize];
+
+	VerifyOrReturnError(devicesPairIndexes, CHIP_ERROR_INTERNAL);
+
+	for (auto i = 0; i < deviceListSize; i++) {
+		indexes[i].SetValue(*devicesPairIndexes[i]);
+		indexPtrs[i] = &indexes[i];
 	}
 
-	if (cumulativeStatus) {
-		dataProviderPtr.release();
-		return CHIP_NO_ERROR;
+	return AddBridgedDevices(devices, dataProvider, deviceListSize, devicesPairIndexes, endpointIds, indexPtrs);
+}
+
+CHIP_ERROR BridgeManager::AddBridgedDevices(MatterBridgedDevice *devices[], BridgedDeviceDataProvider *dataProvider,
+					    uint8_t deviceListSize, uint8_t *devicesPairIndexes[],
+					    uint16_t endpointIds[], chip::Optional<uint8_t> *indexes[])
+{
+	CHIP_ERROR err = AddDevices(devices, dataProvider, deviceListSize, indexes, endpointIds);
+
+	if (err != CHIP_NO_ERROR) {
+		return err;
 	}
-	return CHIP_ERROR_NO_MEMORY;
+
+	for (auto i = 0; i < deviceListSize; i++) {
+		if (!(*indexes[i]).HasValue()) {
+			return CHIP_ERROR_INTERNAL;
```

Comment:
But this not an input argument validation. This argument is actually in/out and in this check we validate if the `AddDevices` method assigned output value, so it can be assigned to the `devicesPairIndexes`. This operation has to succeed, as otherwise we would not return assigned index.

Reply:
Discussed f2f. Currently we have no better idea how to resolve it.